### PR TITLE
Sdf

### PR DESCRIPTION
**Pull request by MatthijsBurgh**
Fixes tue-robotics/tue_robocup#652

### CODE REVIEW

File: include/ed/helpers/msg_conversions.h [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    if (!e.areas().empty())
+    {
+        msg.areas.resize(e.areas().size());
+        int i=0;
+        for (std::map<std::string, geo::ShapeConstPtr>::const_iterator it = e.areas().begin(); it != e.areas().end(); ++it)
+        {
+            ed_msgs::Area area;
+            area.name = it->first;
+
+            geo::CompositeShapeConstPtr composite = boost::dynamic_pointer_cast<const geo::CompositeShape>(it->second);
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply include/ed/helpers/msg_conversions.h]: Ik heb dit besproken met @alberth. Dynamic casten is de enige optie, om een CompositeShape terug te halen van een Shape.

File: include/ed/helpers/msg_conversions.h [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+    sub_area.geometry.dimensions[solid.BOX_Y] = size.y;
+    sub_area.geometry.dimensions[solid.BOX_Z] = size.z;
+}
+
+/**
+ * @brief converting ed::Entity to ed_msgs::EntityInfo message
+ * @param e ed::Entity as input
+ * @param msg filled ed_msgs::EntityInfo message as output
+ */
+void convert(const ed::Entity& e, ed_msgs::EntityInfo& msg) {
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply include/ed/helpers/msg_conversions.h]: But they have different arguments.

File: src/models/xml_shape_parser.cpp [hunk 3]
```diff
@@ -20,10 +20,10 @@ std::vector<double> parseArray(const TiXmlElement* xml_elem)
     std::stringstream stream(txt);
     while(getline(stream, word, ' '))
     {
-        float f = 0;
+        double d = 0;
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply src/models/xml_shape_parser.cpp]: Because the return type is a vector of doubles, so mixing doubles and floats is stupid.

File: src/models/shape_loader.h [hunk 4]
```diff
@@ -4,19 +4,37 @@
...
 
 namespace models
 {
 
+enum ModelOrFile{
+    MODEL,
+    FILE
+};
+
+std::vector<std::string> split(std::string& strToSplit, char delimeter);
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply src/models/shape_loader.h]: Ik heb gegoogled, dit leek de beste optie.

File: tools/view_model.cpp [hunk 5]
```diff
@@ -465,25 +457,17 @@ int main(int argc, char **argv)
...
             else
-                cam_pitch = (int)(cam_pitch / 1.57 - 0.5) * 1.57;
+                cam_pitch = std::round(cam_pitch / 1.57 - 0.51) * 1.57;
 
-            // Snap yaw to 90 degrees
-            if (cam_yaw > 0)
-                cam_yaw = (int)(cam_yaw / 1.57 + 0.5) * 1.57;
-            else
-                cam_yaw = (int)(cam_yaw / 1.57 - 0.5) * 1.57;
+//            // Snap yaw to 90 degrees
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply tools/view_model.cpp]: It was not working properly. Therefore I fixed it. Why the 0.51? I think the 0,5 caused some trouble.

File: include/ed/models/model_loader.h [hunk 6]
```diff
@@ -22,15 +22,19 @@ class ModelLoader
...
+    bool create(const UUID& id, const std::string& type, UpdateRequest& req, std::stringstream& error, const bool allow_sdf=false);
 
     bool create(const tue::config::DataConstPointer& data, UpdateRequest& req, std::stringstream& error);
 
     bool create(const tue::config::DataConstPointer& data, const UUID& id_opt, const UUID& parent_id,
                 UpdateRequest& req, std::stringstream& error, const std::string& model_path = "",
                 const geo::Pose3D& pose_offset = geo::Pose3D::identity());
 
-    bool exists(const std::string& type) const;
+    bool createSDF(const tue::config::DataConstPointer& data, const UUID& parent_id, const geo::Pose3D& parent_pose,
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply include/ed/models/model_loader.h]: Fixed

File: src/convex_hull_calc.cpp [hunk 7]
```diff
@@ -10,6 +10,14 @@ namespace convex_hull
 
 // ----------------------------------------------------------------------------------------------------
 
+/**
+ * @brief create fill a ConvexHull a put its origin in the middle of the Convexhull
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply src/convex_hull_calc.cpp]: fixed

File: include/ed/entity.h [hunk 8]
```diff
@@ -52,6 +54,10 @@ class Entity
     inline geo::ShapeConstPtr shape() const { return shape_; }
     void setShape(const geo::ShapeConstPtr& shape);
 
+    inline std::map<std::string, geo::ShapeConstPtr> volumes() const { return volumes_; }
```
Comment (alberth) [inline include/ed/entity.h]: Given that we also get `addVolume` and `removeVolume`, I guess this should return a `const` map, probably as reference? (unless multiple threads change this? not entirely sure)

File: include/ed/helpers/msg_conversions.h [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+        msg.z_max = convex_hull.z_max;
+
+        msg.convex_hull.resize(convex_hull.points.size());
+        for(unsigned int i = 0; i < msg.convex_hull.size(); ++i)
+        {
+            msg.convex_hull[i].x = convex_hull.points[i].x;
+            msg.convex_hull[i].y = convex_hull.points[i].y;
+            msg.convex_hull[i].z = 0;
+        }
+    }
```
Comment (alberth) [inline include/ed/helpers/msg_conversions.h]: You should initialize the `msg` fields also when the hull has no points I think, or you cannot reliably read the message to see there is no hull.

File: include/ed/helpers/msg_conversions.h [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+    msg.has_shape = e.shape() ? true : false;
```
Comment (alberth) [inline include/ed/helpers/msg_conversions.h]:
```suggestion
    msg.has_shape = (e.shape() != nullptr);
```

File: include/ed/helpers/msg_conversions.h [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+    if (!e.data().empty())
+    {
+        std::stringstream ss;
+        tue::config::YAMLEmitter emitter;
+        emitter.emit(e.data(), ss);
+
+        msg.data = ss.str();
+    }
+
+    if (!e.volumes().empty())
```
Comment (alberth) [inline include/ed/helpers/msg_conversions.h]: A reference thus :)

File: include/ed/models/model_loader.h [hunk 12]
```diff
@@ -22,15 +22,63 @@ class ModelLoader
...
+                   const UUID& id_override, const boost::shared_ptr<const geo::Pose3D> pose_override,
+                   UpdateRequest& req, std::stringstream& error);
+
+    /**
+     * @brief exists Check of a model of type 'type' exist
+     * @param type model type
+     * @return bool, which indicates of type exist
+     */
+    bool exists(std::string type) const;
```
Comment (alberth) [inline include/ed/models/model_loader.h]: I'd prefer `has_model_of_type`, or at least something more precise than `exists`, as lots of things exist in a loader. Not sure `ModelLoader` should have this, as a loader loads stuff rather than store it.

File: include/ed/update_request.h [hunk 13]
```diff
@@ -51,6 +51,30 @@ class UpdateRequest
     void setShape(const UUID& id, const geo::ShapeConstPtr& shape) { shapes[id] = shape; flagUpdated(id); }
 
 
+    //VOLUMES
+
+    std::map<UUID, std::map<std::string, geo::ShapeConstPtr> > volumes_added;
```
Comment (alberth) [inline include/ed/update_request.h]: No `_` at the end?

Also, would be nice if you explain the idea, as I don't understand why you track added volumes and removed volumes.

File: include/ed/update_request.h [hunk 14]
```diff
@@ -51,6 +51,30 @@ class UpdateRequest
...
+        else
+        {
+            std::map<std::string, geo::ShapeConstPtr> volume_map;
+            volume_map[Volume_name] = Volume_shape;
+            volumes_added[id] = volume_map;
+        }
+        flagUpdated(id);
+    }
+
+    std::map<UUID, std::set<std::string> > volumes_removed;
```
Comment (alberth) [inline include/ed/update_request.h]: Same for this variable

File: src/ed.cpp [hunk 15]
```diff
@@ -193,6 +142,7 @@ bool srvUpdate(ed_msgs::UpdateSrv::Request& req, ed_msgs::UpdateSrv::Response& r
                     if (!r.readValue("name", prop_name))
                         continue;
 
+                    // ToDo: is this thread safe?
```
Comment (alberth) [inline src/ed.cpp]: Shouldn't this get done then? same at line 202

File: src/ed.cpp [hunk 16]
```diff
@@ -239,8 +189,8 @@ bool srvUpdate(ed_msgs::UpdateSrv::Request& req, ed_msgs::UpdateSrv::Response& r
 
 bool srvQuery(ed_msgs::Query::Request& req, ed_msgs::Query::Response& res)
 {
-    tue::Timer timer;
-    timer.start();
+//    tue::Timer timer;
+//    timer.start();
```
Comment (alberth) [inline src/ed.cpp]: remove these comments (or is there a policy to leave these?)

File: src/ed.cpp [hunk 17]
```diff
@@ -405,10 +359,11 @@ bool srvSimpleQuery(ed_msgs::SimpleQuery::Request& req, ed_msgs::SimpleQuery::Re
     geo::Vector3 center_point;
     geo::convert(req.center_point, center_point);
 
-    for(ed::WorldModel::const_iterator it = ed_wm->world_model()->begin(); it != ed_wm->world_model()->end(); ++it)
-    {
-//        std::cout << it->first << std::endl;
+    // Make a copy of the WM, to keep it thead safe
+    ed::WorldModel wm = *ed_wm->world_model();
 
+    for(ed::WorldModel::const_iterator it = wm.begin(); it != wm.end(); ++it)
```
Comment (alberth) [inline src/ed.cpp]: Since line 363 is intended for this loop, I'd remove line 364 here

File: src/ed.cpp [hunk 18]
```diff
@@ -249,13 +199,17 @@ bool srvQuery(ed_msgs::Query::Request& req, ed_msgs::Query::Response& res)
...
         const ed::PropertyKeyDBEntry* entry = ed_wm->getPropertyKeyDBEntry(*it);
         if (entry)
             property_idxs.push_back(entry->idx);
     }
 
-    const std::vector<unsigned long>& entity_revs = ed_wm->world_model()->entity_revisions();
-    const std::vector<ed::EntityConstPtr>&  entities = ed_wm->world_model()->entities();
+    // Make a copy of the WM, to keep it thead safe
+    ed::WorldModel wm = *ed_wm->world_model();
+
```
Comment (alberth) [inline src/ed.cpp]: I'd remove this line, as 209 is directly intended for line 211 and further

File: src/entity.cpp [hunk 19]
```diff
@@ -97,7 +97,13 @@ void Entity::updateConvexHullFromShape()
     std::vector<geo::Vec2f> points(vertices.size());
     for(unsigned int i = 0; i < vertices.size(); ++i)
     {
-        geo::Vector3 p_MAP = pose_ * vertices[i];
+//        geo::Vector3 p_MAP = pose_ * vertices[i];
+        // old implementation, this is correct, but gives the wrong result with the rest of the code
+        // Because it is too much work for now to change that. So therefore ignoring rotation.
```
Comment (alberth) [inline src/entity.cpp]: How nice, when is now not now?

File: src/models/model_loader.cpp [hunk 20]
```diff
@@ -20,18 +24,56 @@ namespace ed
...
 {
 
+bool readSDFGeometry(tue::config::Reader r, geo::CompositeShapePtr& composite, std::stringstream& error, geo::Pose3D pose_offset = geo::Pose3D::identity())
+{
+    geo::Pose3D pose = geo::Pose3D::identity();
+    readPose(r, pose);
+    pose = pose_offset * pose;
+    if (r.readGroup("geometry"))
+    {
+        std::map<std::string, geo::ShapePtr> dummy_shape_cache;
```
[the remark this answers is not in the document]
Comment (alberth) [reply src/models/model_loader.cpp]: Allow giving a `nullptr` as cache to skip caching?

File: src/models/model_loader.cpp [hunk 21]
```diff
@@ -20,17 +24,58 @@ namespace ed
...
-    const char * mpath = ::getenv("ED_MODEL_PATH");
+    const char * edmpath = ::getenv("ED_MODEL_PATH");
+    if (edmpath)
+    {
+        std::stringstream ss(edmpath);
+        std::string item;
+        while (std::getline(ss, item, ':'))
+            if (!item.empty())
+                ed_model_paths_.push_back(item);
+    }
```
Comment (alberth) [inline src/models/model_loader.cpp]: Refactor into a function, and use it 3 times

File: src/models/model_loader.cpp [hunk 22]
```diff
@@ -134,15 +289,33 @@ bool ModelLoader::exists(const std::string& type) const
...
+    {
+        sdf = false;
+        data = loadModelData(type, types, error);
+        if (data.empty())
+            return false;
+    }
 
-    if (!create(data, id, "", req, error))
-        return false;
+    // First to get data before trying to create models
```
Comment (alberth) [inline src/models/model_loader.cpp]:
```suggestion
    // First try to get data before trying to create models.
```
(likely)

File: include/ed/helpers/msg_conversions.h (hunk 9 above)
Comment (MatthijsBurgh) [reply include/ed/helpers/msg_conversions.h]: This is done by the constructor of the msg.

File: src/models/model_loader.cpp [hunk 23]
```diff
@@ -134,15 +289,33 @@ bool ModelLoader::exists(const std::string& type) const
...
+    if (sdf)
+    {
+        if (!createSDF(data, "", geo::Pose3D::identity(), id, boost::shared_ptr<const geo::Pose3D>(), req, error))
+            return false;
+    }
+    else
+    {
+        if (!create(data, id, "", req, error))
```
Comment (alberth) [inline src/models/model_loader.cpp]: Quite inconsistent, `create` (without saying what you create) versus `createSDF`.

File: src/models/model_loader.cpp [hunk 24]
```diff
@@ -212,28 +386,10 @@ bool ModelLoader::create(const tue::config::DataConstPointer& data, const UUID&
...
-        // Set rotation
-        pose.R.setRPY(roll, pitch, yaw);
-
-        r.endGroup();
-    }
+    geo::Pose3D pose = geo::Pose3D::identity();
+    if (!ed::models::readPose(r, pose))
+        error << "[ed::models::create] No pose, while reading model: '" << id << "'" << std::endl;
 
     pose = pose_offset * pose;
```
Comment (alberth) [inline src/models/model_loader.cpp]: Continuing computing with poses even if you cannot read it?

File: include/ed/entity.h (hunk 8 above)
Comment (MatthijsBurgh) [reply include/ed/entity.h]: There will never be two threads writing to the same entity instance. It is better if it is a const map

File: src/models/shape_loader.cpp [hunk 25]
```diff
@@ -15,13 +14,134 @@
...
+
+    static std::vector<std::string> model_paths;
+    static std::vector<std::string> file_paths;
+
+    if (model_paths.empty() && file_paths.empty())
+    {
+        std::string item;
+        std::stringstream ssm(mpath);
+        while (std::getline(ssm, item, ':'))
+            model_paths.push_back(item);
```
Comment (alberth) [inline src/models/shape_loader.cpp]: More environment paths getting split

File: include/ed/models/model_loader.h (hunk 12 above)
Comment (MatthijsBurgh) [reply include/ed/models/model_loader.h]: This functionality already existed XD before this PR. So out of scope of this PR.

File: include/ed/update_request.h (hunk 13 above)
Comment (MatthijsBurgh) [reply include/ed/update_request.h]: Because these need to be added/removed. Just like type. No underscore used in update_request at all. So out of this scope.

File: include/ed/update_request.h (hunk 14 above)
Comment (MatthijsBurgh) [reply include/ed/update_request.h]: No underscore used in update_request at all. So out of this scope.

File: src/models/shape_loader.cpp [hunk 26]
```diff
@@ -113,7 +250,7 @@ geo::ShapePtr getHeightMapShape(const std::string& image_filename, const geo::Ve
                 std::vector<geo::Vec2i> points, line_starts;
                 findContours(image, geo::Vec2i(x, y), 0, points, line_starts, contour_map);
 
-                unsigned int num_points = points.size();
+                unsigned int num_points = (unsigned int) points.size();
```
Comment (alberth) [inline src/models/shape_loader.cpp]: Normally you'd use `std::size_t` type for sizes and counts rather than an integer.

File: src/models/shape_loader.cpp [hunk 27]
```diff
@@ -236,7 +373,72 @@ geo::ShapePtr getHeightMapShape(const std::string& image_filename, const geo::Ve
...
+    }
+
+    return getHeightMapShape(image_orig, pos, size, inverted, error);
+}
+
+
+// ----------------------------------------------------------------------------------------------------
+
+/**
+ * @brief getHeightMapShape convert grayscale image in a heigtmap mesh
```
[the remark this answers is not in the document]
Comment (alberth) [reply src/models/shape_loader.cpp]:
```suggestion
 * @brief getHeightMapShape convert grayscale image in a heightmap mesh
```

File: src/models/shape_loader.cpp [hunk 28]
```diff
@@ -517,68 +916,124 @@ geo::ShapePtr loadShape(const std::string& model_path, tue::config::Reader cfg,
...
+                geo::Vec2& p = points.back();
+                std::vector<std::string> point_vector = split(point_string, ' ');
+                p.x = std::stod(point_vector[0]);
+                p.y = std::stod(point_vector[1]);
+            }
+            else
+                cfg.value("height", height);
         }
         cfg.endArray();
+        if (height > 0 && std::abs<double>(height) >= 0.01)
```
Comment (alberth) [inline src/models/shape_loader.cpp]: Why do you perform an `abs` on a known positive number?

File: src/ed.cpp (hunk 17 above)
Comment (MatthijsBurgh) [reply src/ed.cpp]: Done

File: src/models/shape_loader.cpp [hunk 29]
```diff
@@ -517,68 +916,124 @@ geo::ShapePtr loadShape(const std::string& model_path, tue::config::Reader cfg,
...
+            if (mesh_path.exists())
+                shape = geo::Importer::readMeshFile(mesh_path.string());
+            else
+                error << "[ED::MODELS::LOADSHAPE] Mesh File: '" << mesh_path.string() << "' doesn't exist." << std::endl;
+        }
+        else
+            error << "[ED::MODELS::LOADSHAPE] No uri found for mesh." << std::endl;
+
+        std::string dummy;
+        if (cfg.value("submesh", dummy))
```
Comment (alberth) [inline src/models/shape_loader.cpp]: No method to query if a key with a known name exists?

File: src/models/shape_loader.cpp [hunk 30]
```diff
@@ -517,68 +916,124 @@ geo::ShapePtr loadShape(const std::string& model_path, tue::config::Reader cfg,
...
-                && cfg.value("height", height))
+        geo::Vec3 size;
+        if ((cfg.value("image", image_filename) && !image_filename.empty() && cfg.value("resolution", resolution) &&
+             cfg.value("height", height))) // ED YAML ONLY
         {
             std::string image_filename_full = image_filename;
 //            if (image_filename[0] == '/')
 //                image_filename_full = image_filename;
 //            else
 //                image_filename_full = model_path + "/" + image_filename;
```
Comment (alberth) [inline src/models/shape_loader.cpp]: Remove comments

File: src/models/shape_loader.cpp [hunk 31]
```diff
@@ -517,68 +916,124 @@ geo::ShapePtr loadShape(const std::string& model_path, tue::config::Reader cfg,
...
-                readPose(cfg, pose);
-                cfg.endGroup();
-            }
+            readPose(cfg, pose);
+        }
+        else if(cfg.value("uri", image_filename) && readVec3Group(cfg, size, "size")) // SDF ONLY
+        {
+            image_filename = getUriPath(image_filename);
+
+            // by default center should be in the middle.
```
Comment (alberth) [inline src/models/shape_loader.cpp]: And you make it so, just say "is in the middle" ?

File: src/models/shape_loader.h [hunk 32]
```diff
@@ -1,22 +1,48 @@
...
+/**
+ * @brief The ModelOrFile enum This is used to determine the URI type in SDF.
+ */
+enum ModelOrFile
+{
+    MODEL,
+    FILE
+};
+
+std::vector<std::string> split(std::string& strToSplit, char delimeter);
```
Comment (alberth) [inline src/models/shape_loader.h]: Use on paths?
Have a `const` for `strToSplit`?

File: src/models/xml_shape_parser.cpp [hunk 33]
```diff
@@ -20,10 +20,10 @@ std::vector<double> parseArray(const TiXmlElement* xml_elem)
...
         std::istringstream istr(word);
-        istr >> f;
-        v.push_back(f);
+        istr >> d;
+        v.push_back(d);
```
Comment (alberth) [inline src/models/xml_shape_parser.cpp]: Still a horrible name :)

File: src/models/xml_shape_parser.cpp [hunk 34]
```diff
@@ -119,8 +119,8 @@ geo::ShapePtr parseXMLShape(const std::string& filename, std::string& error)
     }
 
     error = s_error.str();
-    if (error.empty())
-        return shape;
+    if (!error.empty())
+        return geo::ShapePtr();
 
-    return geo::ShapePtr();
+    return shape;
```
Comment (alberth) [inline src/models/xml_shape_parser.cpp]: Could be a `? .. : ..` but fair enough.

File: src/ed.cpp (hunk 18 above)
Comment (MatthijsBurgh) [reply src/ed.cpp]: Done

File: src/ed.cpp (hunk 16 above)
Comment (MatthijsBurgh) [reply src/ed.cpp]: Done

File: tools/heightmap_to_mesh.cpp [hunk 35]
```diff
@@ -0,0 +1,78 @@
...
+//            geo::Vector3 p3(it->p3_);
+//            p3.z = 0;
+//            int i1 = mesh.addPoint(geo::Vector3(it->p1_.x, it->p1_.y, 0.));
+//            int i2 = mesh.addPoint(geo::Vector3(it->p2_.x, it->p2_.y, 0.));
+//            int i3 = mesh.addPoint(geo::Vector3(it->p3_.x, it->p3_.y, 0.));
+//            mesh.addTriangle(i3, i2, i1);
+//        }
+//    }
+//    mesh.filterOverlappingVertices();
+//    shape->setMesh(mesh);
```
Comment (alberth) [inline tools/heightmap_to_mesh.cpp]: doesn't look like it should be here

File: src/ed.cpp (hunk 15 above)
Comment (MatthijsBurgh) [reply src/ed.cpp]: This is out of the scope. Of this PR. We haven't been experiencing any issues related these, but I added these comments for the future, because there are some concerns there.

File: tools/view_model.cpp [hunk 36]
```diff
@@ -206,7 +220,7 @@ void CallBackFunc(int event, int x, int y, int flags, void* userdata)
         }
         else if (flags & cv::EVENT_FLAG_MBUTTON)
         {
-            cam_dist += cam_dist * dy * 0.002;
+            cam_dist += cam_dist * dy * 0.003;
```
Comment (alberth) [inline tools/view_model.cpp]: Any chance of explaining what the magic constant is about? Or better, make a constant with a proper name for it?

File: src/models/model_loader.cpp [hunk 37]
```diff
@@ -56,36 +107,95 @@ std::string ModelLoader::getModelPath(const std::string& type) const
...
+ModelLoader::ModelData ModelLoader::readModelCache(std::string type) const
+{
+    std::map<std::string, ModelData>::const_iterator it = model_cache_.find(type);
     if (it != model_cache_.end())
+        return it->second;
+
+    tue::config::DataConstPointer data;
+    std::vector<std::string> types;
+    return  ModelData(data, types);
+}
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply src/models/model_loader.cpp]: Out of the scope of this PR.

File: src/models/model_loader.cpp [hunk 38]
```diff
@@ -108,24 +218,69 @@ tue::config::DataConstPointer ModelLoader::loadModelData(const std::string& type
...
-    model_cache_[type] = ModelData(data, types);
+    model_cache_[parsed_uri+"_sdf"] = ModelData(data, std::vector<std::string>());
 
     return data;
 }
 
 // ----------------------------------------------------------------------------------------------------
 
-bool ModelLoader::exists(const std::string& type) const
+bool ModelLoader::exists(std::string type) const
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply src/models/model_loader.cpp]: Fixed

File: tools/view_model.cpp [hunk 39]
```diff
@@ -284,27 +295,30 @@ int main(int argc, char **argv)
...
+    info_msg << "    p - snap pitch" << std::endl;
+    info_msg << "    q - quit" << std::endl;
 
-    std::cout << std::endl;
-    std::cout << "Keys:" << std::endl;
-    std::cout << "    r - reload model" << std::endl;
-    std::cout << "    a - show / hide model areas" << std::endl;
-    std::cout << "    c - circle rotate" << std::endl;
-    std::cout << "    q - quit" << std::endl;
+    std::cout << info_msg.str();
```
Comment (alberth) [inline tools/view_model.cpp]: Why not dump a multi-line C string in to cout?

```` cpp
cout << "blah\n"
            "more blah\n"
            "etc\n";
````
Since 2 strings after each other simply get concatenated.

File: src/models/model_loader.cpp [hunk 40]
```diff
@@ -20,18 +24,56 @@ namespace ed
 namespace models
 {
 
+bool readSDFGeometry(tue::config::Reader r, geo::CompositeShapePtr& composite, std::stringstream& error, geo::Pose3D pose_offset = geo::Pose3D::identity())
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply src/models/model_loader.cpp]: Fixed

File: tools/view_model.cpp [hunk 41]
```diff
@@ -465,25 +453,17 @@ int main(int argc, char **argv)
...
-            // Snap yaw to 90 degrees
-            if (cam_yaw > 0)
-                cam_yaw = (int)(cam_yaw / 1.57 + 0.5) * 1.57;
-            else
-                cam_yaw = (int)(cam_yaw / 1.57 - 0.5) * 1.57;
+//            // Snap yaw to 90 degrees
+//            if (cam_yaw < 0)
+//                cam_yaw = std::round(cam_yaw / 1.57 + 0.51) * 1.57;
+//            else
+//                cam_yaw = std::round(cam_yaw / 1.57 - 0.51) * 1.57;
```
Comment (alberth) [inline tools/view_model.cpp]: Explain why it's disabled?, or remove it?

File: src/models/model_loader.cpp [hunk 42]
```diff
@@ -20,18 +24,56 @@ namespace ed
...
+        if (sub_shape)
+        {
+            if (!composite)
+                composite.reset(new geo::CompositeShape);
+            composite->addShape(*sub_shape, pose);
+        }
+        r.endGroup();
+        return true;
+    }
+    else
```
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply src/models/model_loader.cpp]: Done

File: src/models/model_loader.cpp (hunk 24 above)
Comment (MatthijsBurgh) [reply src/models/model_loader.cpp]: Not having a pose isn't really an error.

File: src/models/shape_loader.cpp (hunk 26 above)
Comment (MatthijsBurgh) [reply src/models/shape_loader.cpp]: That would be better, but not used in this file. So for consistency keeping it as it is.

File: src/models/shape_loader.cpp (hunk 28 above)
Comment (MatthijsBurgh) [reply src/models/shape_loader.cpp]: Fixed.

File: src/models/shape_loader.cpp (hunk 29 above)
Comment (MatthijsBurgh) [reply src/models/shape_loader.cpp]: No these only exist for Map and Array childs. Not for value.

File: src/models/shape_loader.cpp (hunk 31 above)
Comment (MatthijsBurgh) [reply src/models/shape_loader.cpp]: Fixed

File: src/models/xml_shape_parser.cpp (hunk 33 above)
Comment (MatthijsBurgh) [reply src/models/xml_shape_parser.cpp]: Yep, but not high prio.

File: src/models/xml_shape_parser.cpp (hunk 34 above)
Comment (MatthijsBurgh) [reply src/models/xml_shape_parser.cpp]: Nope, with scope indeed better.

Comment (MatthijsBurgh) [reply src/models/xml_shape_parser.cpp]: Could indeed be confusing

File: tools/heightmap_to_mesh.cpp (hunk 35 above)
Comment (MatthijsBurgh) [reply tools/heightmap_to_mesh.cpp]: Removed

File: tools/view_model.cpp (hunk 36 above)
Comment (MatthijsBurgh) [reply tools/view_model.cpp]: I don't know :)

File: tools/view_model.cpp (hunk 39 above)
Comment (MatthijsBurgh) [reply tools/view_model.cpp]: I don't know. This is already better then the previous implementation.

File: tools/view_model.cpp (hunk 41 above)
Comment (MatthijsBurgh) [reply tools/view_model.cpp]: removed

File: src/models/shape_loader.h (hunk 32 above)
Comment (MatthijsBurgh) [reply src/models/shape_loader.h]: Done

File: src/models/shape_loader.cpp (hunk 30 above)
Comment (MatthijsBurgh) [reply src/models/shape_loader.cpp]: I would like to keep it, as it was already there.

File: src/models/shape_loader.cpp (hunk 25 above)
Comment (MatthijsBurgh) [reply src/models/shape_loader.cpp]: Fixed

File: src/models/model_loader.cpp (hunk 22 above)
Comment (MatthijsBurgh) [reply src/models/model_loader.cpp]: Fixed

File: src/models/model_loader.cpp (hunk 38 above)
[the remark this answers is not in the document]
Comment (MatthijsBurgh) [reply src/models/model_loader.cpp]: Fixed

File: src/models/model_loader.cpp (hunk 21 above)
Comment (MatthijsBurgh) [reply src/models/model_loader.cpp]: fixed

File: include/ed/helpers/msg_conversions.h (hunk 9 above)
Comment (alberth) [reply include/ed/helpers/msg_conversions.h]: but you're not constructing a `msg` here, so there is no way of knowing it has the proper values.

File: include/ed/update_request.h (hunk 13 above)
Comment (alberth) [reply include/ed/update_request.h]: I don't understand your explanation, but fair enough for now.

File: src/ed.cpp (hunk 15 above)
Comment (alberth) [reply src/ed.cpp]: "concerns" is an under-statement I think :)

File: src/models/model_loader.cpp (hunk 23 above)
Comment (alberth) [reply src/models/model_loader.cpp]: Getting name problems typically means you're trying to push too much different functionality into a single name. Perhaps is also related to comment by Janno about names being too generic.

Fair enough for now though, let's leave this for tomorrow.

**Comment by alberth** [comment]
Nice, it doesn't show the new comments :(

Anyway, only proper initialization in `msg_conversions.h`, function `void convert(const ed::Entity& e, ed_msgs::EntityInfo& msg)`, since you don't know if `msg` has not been used before,

File: include/ed/helpers/msg_conversions.h (hunk 9 above)
Comment (MatthijsBurgh) [reply include/ed/helpers/msg_conversions.h]: This is the output of the service call to `simple_query`:
```
entities: 
  - 
    id: "exit_1_rips"
    type: "waypoint"
    types: [waypoint]
    existence_probability: 1.0
    last_update_time: 
      secs: 0
      nsecs:         0
    creation_time: 
      secs: 0
      nsecs:         0
    has_shape: False
    has_pose: True
    pose: 
      position: 
        x: -1.48
        y: -0.031
        z: 0.0
      orientation: 
        x: 0.0
        y: 0.0
        z: 0.999999998927
        w: 4.632679488e-05
    data: ''
    flags: []
    convex_hull: []
    z_min: 0.0
    z_max: 0.0
    volumes: []
```
`convert` is called with a reference of the msg. So the message is constructed somewhere else.
The constructor of the message is:
```cpp
EntityInfo_()
    : id()
    , type()
    , types()
    , existence_probability(0.0)
    , last_update_time()
    , creation_time()
    , has_shape(false)
    , has_pose(false)
    , pose()
    , data()
    , flags()
    , convex_hull()
    , z_min(0.0)
    , z_max(0.0)
    , volumes()  {
    }
```
So I still think the message is initialized correct. So the convex_hull will be an empty vector.
